src/lib.rs

```diff
@@ -9,11 +9,151 @@
+                registry.args_hidden = Some(Vec::new());
+                plugin_registrar(&mut registry);
+
+                let rustc_plugin::registry::Registry { early_lint_passes, late_lint_passes, lint_groups, llvm_passes, attributes, mir_passes, .. } = registry;
+                let sess = &state.session;
+                let mut ls = sess.lint_store.borrow_mut();
+                for pass in early_lint_passes {
+                    ls.register_early_pass(Some(sess), true, pass);
+                }
+                for pass in late_lint_passes {
+                    ls.register_late_pass(Some(sess), true, pass);
+                }
+
+                for (name, to) in lint_groups {
+                    ls.register_group(Some(sess), true, name, to);
+                }
+
+                sess.plugin_llvm_passes.borrow_mut().extend(llvm_passes);
+                sess.mir_passes.borrow_mut().extend(mir_passes);
+                sess.plugin_attributes.borrow_mut().extend(attributes);
+            }
+            old(state);
+        });
+
+        control
+    }
+}
+
+use std::path::Path;
+
+pub fn main() {
+    use std::env;
+
+    if env::var("DOGFOOD").map(|_| true).unwrap_or(false) {
+        return;
+    }
+
+    let dep_path = env::current_dir().expect("current dir is not readable").join("target").join("debug").join("deps");
+    let sys_root = match (option_env!("MULTIRUST_HOME"), option_env!("MULTIRUST_TOOLCHAIN")) {
+        (Some(home), Some(toolchain)) => format!("{}/toolchains/{}", home, toolchain),
+        _ => option_env!("SYSROOT").expect("need to specify SYSROOT env var during clippy compilation").to_owned(),
+    };
+
+    if env::var("AUTOBOTS").map(|s| s == "ROLLOUT").unwrap_or(false) {
+        let args: Vec<String> = if env::args().any(|s| s == "--sysroot") {
+            env::args().collect()
+        } else {
+            env::args().chain(Some("--sysroot".to_owned())).chain(Some(sys_root)).collect()
+        };
+        rustc_driver::run_compiler(&args, &mut ClippyCompilerCalls::new());
+    } else {
+        let args = wrap_args(env::args(), dep_path, sys_root);
+        let path = std::env::current_exe().expect("current executable path invalid");
+        let run = std::process::Command::new("cargo")
+            .args(&args)
+            .env("RUSTC", path)
+            .env("AUTOBOTS", "ROLLOUT")
+            .spawn().expect("could not run cargo")
+            .wait().expect("failed to wait for cargo?")
+            .success();
+        assert!(run, "cargo rustc failed");
+    }
+}
+
+fn wrap_args<T, I, P>(it: I, dep_path: P, sysroot: String) -> Vec<String>
+    where T: AsRef<str>,
+          I: IntoIterator<Item=T>,
+          P: AsRef<Path> {
+
+    let old_args = it.into_iter();
+    let mut args = vec!["rustc".to_owned()];
+
+    let mut found_dashes = false;
+    for arg in old_args.skip(2) {
```

Comment:
it can be invoked as cargo clippy and cargo-clippy, so perhaps we should skip conditionally